### DISCUSSION TITLE
Latency when using a Raspberry Pi Zero

### DISCUSSION THREAD

**Discussion by DutchBigEd** [probonopd/MiniDexed#433] [Q&A] [answered]
Ok, I got me a Raspberry Pi Zero V1.2
The idea is to build it into an old midi keyboard and **I need only one instance**, so my guess is the Zero will do just fine...
I have connected an Audio Hat , a Midi input on the RX, a LCD screen 2x16 and three buttons on GPIO 9, 10 and 11.
I have good, clear sound when connecting a midi keyboard, I can change voices and banks using midi-ox or surf the banks and voices with the buttons, so far so good.
Also the display shows what it should, but...
Is there a way to minimise the latency?

**Comment by probonopd** [comment, marked as answer]
> Is there a way to minimise the latency?

First step would be to _measuure_ it. If you are already using Serial MIDI and a i2s DAC, then I don't know how it could be further minimized without switching to a more powerful Raspberry Pi model.

**Reply by DutchBigEd** [reply]
![20230201_084222](https://user-images.githubusercontent.com/17274454/215985233-40abba28-dc91-4875-b551-53643ff3bfd2.jpg)
I have chosen a triangle2 from the dx7-2 bank.
Left cursor is note-on and right cursor is the audio signal.
The scope is set to 20msec/div, so the latency is about 40msec. (sometimes a little less, sometimes a little more)
Removing usbspeed=full from cmdline.txt has no visible effect.

**Comment by Banana71** [comment]
My Raspberry Pi Model 2B with Classic MIDI runs much better if I remove usbspeed=full in the cmdline.txt file.  or   //usbspeed=full

**Reply by DutchBigEd** [reply]
I'm not using usb-midi and disabled midi thru.

**Reply by Banana71** [reply]
yes, that's exactly why i would change to //usbspeed=full in cmdline.txt

**Comment by DutchBigEd** [comment]
![keyboard minidexed](https://user-images.githubusercontent.com/17274454/216029275-ce8c9ca1-5746-45ad-b96b-fdbbce5121a7.jpg)
One instance minidexed DX7

**Comment by donluca** [comment]
Without creating a new thread... does Minidexed support 1000Hz USB polling?

It is an enhancement which (to the best of my knowledge) was first proposed by Razer for its gaming peripherals (mice and keyboards) to poll USB faster and reduce latency and, since then, it has become the standard used by gamers all over the world, with measurable improvement over the default 125Hz (IIRC).

I admit I have never used it with music peripherals such as keyboards and such, but I wonder if Minidexed could benefit from it.

**Reply by probonopd** [reply]
Good point. @rsta what do you think about it? Maybe the polling rate for USB could be made configurable in Circle.

**Comment by Gustavomurta** [comment]
Donluca, 

My configuration is:
- Raspberry Pi Zero (old model)
- Midi KBD : Akai Mpk mini (USB MIDI) 

In my perception, the USB protocol on the mini Dexed already uses a very fast sync.

Logic analyzer screens: 
I2S DIN - brown when key pressed. 

![image](https://github.com/probonopd/MiniDexed/assets/4587366/b181fc05-cffe-44e1-ab4f-29d1df20e18d)





![image](https://github.com/probonopd/MiniDexed/assets/4587366/c43837d1-ccaa-4cd3-aff5-d00a6e077f4e)
